src/cluster/mcl.cpp

```diff
@@ -0,0 +1,443 @@
+	// data.emplace_back(5, 9, 0.25f);
+	// data.emplace_back(6, 0, 0.20f);
+	// data.emplace_back(6, 4, 0.20f);
+	// data.emplace_back(6, 6, 0.25f);
+	// data.emplace_back(6, 9, 0.25f);
+	// data.emplace_back(7, 3, 0.20f);
+	// data.emplace_back(7, 4, 0.20f);
+	// data.emplace_back(7, 7, 0.20f);
+	// data.emplace_back(7, 8, 0.20f);
+	// data.emplace_back(7, 10, 0.20f);
+	// data.emplace_back(8, 3, 0.20f);
+	// data.emplace_back(8, 7, 0.20f);
+	// data.emplace_back(8, 8, 0.20f);
+	// data.emplace_back(8, 10, 0.20f);
+	// data.emplace_back(8, 11, 0.333f);
+	// data.emplace_back(9, 0, 0.20f);
+	// data.emplace_back(9, 5, 0.333f);
+	// data.emplace_back(9, 6, 0.25f);
+	// data.emplace_back(9, 9, 0.25f);
+	// data.emplace_back(10, 3, 0.20f);
+	// data.emplace_back(10, 7, 0.20f);
+	// data.emplace_back(10, 8, 0.20f);
+	// data.emplace_back(10, 10, 0.20f);
+	// data.emplace_back(10, 11, 0.333f);
+	// data.emplace_back(11, 8, 0.20f);
+	// data.emplace_back(11, 10, 0.20f);
+	// data.emplace_back(11, 11, 0.333f);
+	// vector<unordered_set<uint32_t>> cr;
+	// float inf = 2.0;
+	// float exp = 2.0;
+	// if(false){
+	// 	Eigen::SparseMatrix<float> tmp(12,12);
+	// 	tmp.setFromTriplets(data.begin(), data.end());
+	// 	printMatrix(&tmp);
+	// 	cr = markov_process(&tmp, inf, exp);
+	// 	printMatrix(&tmp);
+	// }
+	// else{
+	// 	Eigen::MatrixXf tmp = Eigen::MatrixXf::Zero(12,12);
+	// 	for(Eigen::Triplet<float> const & t : data){
+	// 		tmp(t.row(), t.col()) = t.value();
+	// 	}
+	// 	printMatrix(&tmp);
+	// 	cr = markov_process(&tmp, inf, exp);
+	// 	printMatrix(&tmp);
+	// }
+	// cout << "Found "<<cr.size()<<" clusters"<<endl;
+	// for(auto& set: cr){
+	// 	cout<< "set :";
+	// 	for(auto item : set){
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+	// return;
+
+	const size_t seq_count = db->ref_header.sequences;
+	statistics.reset();
+	config.command = Config::blastp;
+	config.no_self_hits = false;
+	config.output_format = { "clus" };
+	config.query_cover = 80;
+	config.subject_cover = 80;
+	config.algo = 0;
+	config.index_mode = 0;
+	config.freq_sd = 0;
+	config.max_alignments = numeric_limits<uint64_t>::max();
+
+	Workflow::Search::Options opt;
+	opt.db = &(*db);
+	opt.self = true;
+	SparseMatrixStream<float> ms(db->ref_header.sequences);
+	opt.consumer = &ms;
+	opt.db_filter = nullptr;
+	Workflow::Search::run(opt);
+
+	task_timer timer;
+	timer.go("Computing independent components");
+	auto componentInfo = ms.getComponents();
+	vector<vector<uint32_t>> indices = get<0>(componentInfo);
+	vector<vector<Eigen::Triplet<float>>> components = get<1>(componentInfo);
+	uint32_t nComponents = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 0;});
+	uint32_t nComponentsLt1 = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 1;});
+	timer.finish();
+	cout << "Found " << nComponentsLt1 <<" ("<<nComponents<< " incl. singletons) disconnected components" << endl;
```

Comment:
Writing to `cout` should be reserved for output. You can use the `message_stream` object here.